### DISCUSSION TITLE
Snap Your Sector With Deals!

### DISCUSSION THREAD

**Discussion by jennijuju** [filecoin-project/lotus#8141] [Tutorials]
> **For more Snap Deal Documentation - please go to [lotus doc](https://lotus.filecoin.io/docs/storage-providers/snap-deals/), we will merge the tutorial there as well once it's completed!**

# What is Snap Deal?

Filecoin Network [v15 OhSnap upgrade](https://github.com/filecoin-project/community/discussions/74?sort=new#discussioncomment-1922550) will introduce [FIP-0019](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0019.md) - Snap Deals, which is a protocol for updating eligible [CC sectors](https://spec.filecoin.io/#section-glossary.capacity-commitment) with deals with a single message to the filecoin blockchain.

Unlike the current CC upgrade, in which a new sector will be created and the sector will go through the entire sealing pipeline(AP, PC1/2, C1/2 & etc..) to get the deal sealed, snap deals allows a deal to be placed in a CC sector with a shorter, less expensive compute pipeline (UpdateReplica, ProveReplicaUpdate). According to the latest results from the testnest, this reduces deal activation time from 5-6 hours to 1-hour or less.  Of course these results are hardware dependent, let us know how fast you may get data through the snap deals pipeline and into in the Filecoin network [here](https://github.com/filecoin-project/lotus/discussions/8127)!

# Things You Should Know Before You Snap Via Lotus

> Note: Snap Deal is a relatively advanced SP operation. If you aren't familiar with Filecoin, Lotus, or basic storage provider operations and sector/deal life cycles, or how to become a storage provider using lotus in general yet, please visit [Lotus Docs](lotus.filecoin.io) first!

## Say Goodbye to mark-for-upgrade

2 days before the OhSnap upgrade epoch, the [existing upgrading path](https://lotus.filecoin.io/docs/storage-providers/sector-pledging/#upgrading-pledged-sectors) will be deprecated. 
That being said, if you run `lotus-miner sectors mark-for-upgrade` :
- within 2 days before the upgrade, you will get an error message suggesting `OhSnap is coming soon, please use `snap-up` to upgrade your cc sectors after the network v15 upgrade!`.
- after the upgrade epoch, you'll get `classic cc upgrades disabled v15 and beyond, use `snap-up``.

## Configuration

There are a couple configuration worth calling out:

-  in [Dealmaking](https://lotus.filecoin.io/docs/storage-providers/config/#dealmaking-section) session, `ExpectedSealDuration` applies to both Snap Deal process or regular sealing pipeline.
-  in [Sealing](https://lotus.filecoin.io/docs/storage-providers/config/#sealing-section), `MaxWaitDealsSectors`, `MaxSealingSectorsForDeals`, `WaitDealsDelay`, `AlwaysKeepUnsealedCopy` applies to all Deal sectors, including SnapDeal sectors
-  set `AllowReplicaUpdate` and `AllowProveReplicaUpdate` with true/false to allow/disallow corresponding job.
-  `MaxCommitGasFee` and `CommitControl` address are applied to `ProveReplicaUpdate` as well.
  
## Snap Sector States 

All Snaps start with an active CC sector, and under happy path, it will go through the following states (that shows in `lotus-miner sectors list`):
- `SnapDealsWaitDeals`: the CC sector is snapped and is waiting for deals. It will move to the next states if
  - the sector is full with the deals or hits the max amount of the deals one sector may hold
  - `WaitDealsDelay` is hit
- `SnapDealsAddPiece`: same with `AddPiece`
- `SnapDealsPacking`: same with `Packing`
- `UpdateReplica`: move to the next state if all pieces pass validation(size, expiration and so on) and snap deal replica update is generated successfully
  - you may enable this step on a worker by running `lotus-worker run --replica-update=true` or `lotus-worker tasks enable RU`
  - you will see `/.lotusminer/update-cache/s-<minerid>-<secNum>` and  `/.lotusminer/update/s-<minerid>-<secNum>`  file being created during this process
- `ProveReplicaUpdate`: this sealing FSM state does two worker calls
  - `ProveReplicaUpdate1` generates the vanilla proof
  - `ProveReplicaUpdate2` generates the replica update with vanilla proof and proves that snap deals replica was done correctly, then it will move to the next states
    - piece validation will be done before this process starts
    - `ProveReplicaUpdate1` is enabled by default on sealing workers
    - you may enable the second step on a worker by running `lotus-worker run --prove-replica-update2=true` or `lotus-worker tasks enable PR2`
     - using GPU may increase the performance on `ProveReplicaUpdate2`
- `SubmitReplicaUpdate`: submit the `ReplicaUpdate` message to the chain with sector information, including updated sealed sector cid, deal IDs, replica proof and so on along with the collateral(if applicable)
- `ReplicaUpdateWait`: wait for the message land onchain successfully then move to the next state
- `FinalizeReplicaUpdate`: moves sector data to long term storage and cleans up the update cache
- `UpdateActivating` & `ReleaseSectorKey`: clear the cached files safely then your sector will be ready for `Proving` again 

## How to Snap?

It's super simple, run `lotus-miners sectors list | grep CC` to find your all your CC sectors, then run `lotus-miner sectors snap-up <secNum>` - that's it! Now you should see your sector status has turned into `SnapDealsWaitDeals` and ready for deals!

## How to Undo a Snap?

To undo a Snap, run `lotus-miner sectors update-state  --really-do-it=true AbortUpgrade` anytime. Note: this will remove all the deal pieces and turn any sector to CC sector in a `Proving` state. 

You can then snap-up it again as desired.

*CAUTION* Doing this after `SubmitReplicaUpdate` can lead to a sector fault. This command should only be called if you are sure your UpdateReplica message has *not* made it to the chain

**Comment by Shekelme** [comment]
I'm afraid that correct syntax is `lotus-miner sectors abort-upgrade --really-do-it <Sector_No>`
